File: cp_cpp/src/bin/brillianceofwings.cpp
// Created by Ayush Biswas at 2026/04/18 14:41
// https://open.kattis.com/problems/brillianceofwings

/// @head begin
#include <set>
#include <vector>
#include <ranges>
#include <iostream>
#include <algorithm>
/// @head end

/// @code begin
using namespace std;

void sol() {
    int n; cin >> n;
    vector<set<int>> graph1(n + 1);
    for (int i = 1; i < n; i++) {
        int x, y; cin >> x >> y;
        graph1[x].insert(y);
        graph1[y].insert(x);
    }
    vector<set<int>> graph2(n + 1);
    for (int i = 1; i < n; i++) {
        int x, y; cin >> x >> y;
        graph2[x].insert(y);
        graph2[y].insert(x);
    }

    set<int> r;
    int res = 2*(n - 1);
    for (auto [a, b] : views::zip(graph1, graph2)) {
        ranges::set_intersection(a, b, inserter(r, r.begin()));
        res -= (int) r.size();
    }

    cout << res / 2 << endl;
}

int main() {
    cin.tie(nullptr);

    sol();
    return 0;
}
/// @code end

File: cp_rust/src/bin/brillianceofwings.rs
// Created by Ayush Biswas at 2026/05/06 18:34
// https://open.kattis.com/problems/brillianceofwings
use cp_lib::*;

// @code begin
use cpio::*;
use itertools::Itertools;
use std::collections::HashSet;

fn hash(mut l: [usize; 2], m: usize) -> usize {
    l.sort();
    let [x, y] = l;
    x * m + y
}

sol! {
    fn solution(
        n: usize,
        a: [[usize; 2]; n - 1],
        b: [[usize; 2]; n - 1]
    ) -> usize {
        let m = n + 1;
        let a: HashSet<usize> = a
            .into_iter()
            .map(|ai| hash(ai, m))
            .collect();

        let b: HashSet<usize> = b
            .into_iter()
            .map(|ai| hash(ai, m))
            .collect();

        a.symmetric_difference(&b).collect_vec().len() / 2
    }
}
// @code end
